### DISCUSSION TITLE
How to check if OptiX is available

### DISCUSSION THREAD

**Discussion by maxfrei750** [mitsuba-renderer/mitsuba3#360] [Q&A] [answered]
I ran into a similar error as mentioned in #167, but for `libnvoptix.so.1`. 

When loading Mitsuba, I receive:
```
jit_optix_init(): libnvoptix.so.1 could not be loaded -- disabling OptiX backend! Set the DRJIT_LIBOPTIX_PATH environment variable to specify its path.
```

However, if I insist on running with a CUDA variant, I receive `RuntimeError('Could not initialize OptiX!')` (rightly so). Since there is no solution yet (waiting for NVIDIA, see [here](https://forums.developer.nvidia.com/t/optix-on-wsl-windows-subsystem-for-linux/140297/15)), I'd like to catch the error by checking if OptiX is available and setting the Mitsuba variant to LLVM, if it is not. 

Is there a way to check if OptiX is available, using `drjit` or `mitsuba` from Python?

Thanks!

EDIT: I think I found the relevant part of the code (`jitc_optix_check`) in https://github.com/mitsuba-renderer/drjit-core/blob/044f04b54fce4af81fd841352fed594889e6b608/src/optix_api.cpp, but is there any way to call it from Python?

**Comment by njroussel** [comment]
The mechanism to dynamically load shared libraries is mostly handled by `jitc_find_library` in `drjit-core` ([code](https://github.com/mitsuba-renderer/drjit-core/blob/master/src/init.cpp#L604-L687)). Unfortunately, this function is not directly bound to Python, and I can't think of a clever way to manipulate any other existing code to re-use this.

You would need to mimic it's behaviour, which is pretty straightforward. All we're doing is `dlopen("libnvoptix.so.1")`, which you should be able to reproduce with `ctypes`. In Dr\.Jit a failed load of this library deactivates the OptiX backend.

**Reply by maxfrei750** [reply]
Thanks for the quick and extremely helpful reply. I created an answer with an example code.

**Comment by maxfrei750** [comment, marked as answer]
Thanks to the input of @njroussel, I came up with the following function that checks if OptiX is available:

```python
import ctypes

def is_optix_available() -> bool:
    """Check, if OptiX is available."""
    try:
        ctypes.cdll.LoadLibrary("libnvoptix.so.1")
    except OSError:
        return False

    return True
```
